rig/place_and_route/place/sa/c_kernel.py

```diff
@@ -0,0 +1,129 @@
+    kernel requires a compiler it is not included in the standard Rig package.
+    The C kernel can be installed using::
+
+        $ pip install rig_c_sa
+
+    The `rig_c_sa <https://github.com/project-rig/rig_c_sa>`_ package is a
+    Python/C library containing the C components of the algorithm. Installation
+    requires that you have C compiler installed along with `libffi
+    <https://sourceware.org/libffi/>`_ which are readily available in the
+    package managers of real operating systems.
+    """
+
+    def __init__(self, vertices_resources, movable_vertices, fixed_vertices,
+                 initial_placements, nets, machine, random, **kwargs):
+
+        # Seed the random number generator
+        rig_c_sa.srand(random.getrandbits(32))
+
+        # Allocate the basic state for the kernel
+        self.s = rig_c_sa.sa_new(machine.width, machine.height,
+                                 len(machine.chip_resources),
+                                 len(vertices_resources),
+                                 len(nets))
+        assert self.s != ffi.NULL
+
+        # Automatically free SA state on GC of pointer to it
+        self.s = ffi.gc(self.s, rig_c_sa.sa_free)
+
+        # Set basic kernel attributes
+        self.s.has_wrap_around_links = machine.has_wrap_around_links()
+        self.s.num_movable_vertices = len(movable_vertices)
+
+        # Get lookup from vertices to set of nets, suppressing duplicates.
+        vertices_nets = defaultdict(set)
+        for net in nets:
+            for vertex in net:
+                vertices_nets[vertex].add(net)
+
+        # Create all vertices and set their initial positions. Populates a from
+        # Python vertex object to C sa_vertex_t pointer
+        self.vertices_c = {}
+        for i, (vertex, movable) in enumerate(
+                chain(zip(movable_vertices, repeat(True)),
+                      zip(fixed_vertices, repeat(False)))):
+            # Create the vertex
+            v = rig_c_sa.sa_new_vertex(self.s, len(vertices_nets[vertex]))
+            assert v != ffi.NULL
+            self.s.vertices[i] = v
+
+            # Add to lookup
+            self.vertices_c[vertex] = v
+
+            # Set resource consumption
+            for i, resource in enumerate(machine.chip_resources):
+                v.vertex_resources[i] = \
+                    vertices_resources[vertex].get(resource, 0)
+
+            # Add to chip selected by initial placement
+            x, y = initial_placements[vertex]
+            rig_c_sa.sa_add_vertex_to_chip(self.s, v, x, y, movable)
+
+        # Create all nets
+        for i, net in enumerate(nets):
+            # Remove duplicate vertices from the net (e.g. for self-loops only
+            # include the source/sink vertex once).
+            vertices = set(net)
+
+            # Create the net
+            n = rig_c_sa.sa_new_net(self.s, len(vertices))
+            assert n != ffi.NULL
+            self.s.nets[i] = n
+
```

Comment:
I think I'd remove this empty line to put `n.weight ...` more directly under `Create the net`... but this is some major pedantry!